### PR TITLE
JBPM-9484 Improve BPMN -> FEEL integration

### PR DESCRIPTION
**Pull request by tarilabs**
**JIRA**: https://issues.redhat.com/browse/JBPM-9484

**referenced Pull Requests**: none

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluator.java [hunk 1]
```diff
@@ -74,16 +78,17 @@ public Object evaluate(ProcessContext context) throws Exception {
...
         }
         DMNRuntime runtime = ((KieSession) context.getKieRuntime()).getKieRuntime(DMNRuntime.class);
         List<FEELProfile> profiles = (List)((DMNRuntimeImpl) runtime).getProfiles();
         FEEL feel = FEEL.newInstance(runtime.getRootClassLoader(), profiles);
+        FeelReturnValueEvaluatorListener listener = new FeelReturnValueEvaluatorListener();
+        feel.addListener(listener);
         
         Object value = feel.evaluate(expr, variables);
 
+        listener.getEvents().forEach(FeelReturnValueEvaluator::processEvents);
```
Comment (fjtirado) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluator.java:91]: If we want to process all events generated by the listener and do not fail with the first error, I guess we will need something like
StringBuilder errorMessage = new StringBuilder();
for (FeelEvent : listerner.getEvents()) {
    processEvent(event,errorMessage);
}
if (errorMessage.length()>0) { throw new FeelReturnValueEvaluatorException(errorMessage);

and 
private static void process(FEELEvent event, StringBuilder messageBuilder) {

....
 case ERROR:
                StringBuilder messageBuilder = new StringBuilder(event.getSeverity().toString()).append(" ").append(event.getMessage());
                if (event.getOffendingSymbol() != null) {
                    messageBuilder.append(" ( offending symbol: '").append(event.getOffendingSymbol()).append("' )");
                }
                if (event.getSourceException() != null) {
                    messageBuilder.append("  ").append(event.getSourceException().getMessage());
                }
                 LOG.error("Event: {} Error: {}", event, messageBuilder);
.....
}                  
    


}
     
}

Comment (fjtirado) [reply jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluator.java:91]: If we want to fail with the first error, then I will change the listener to generate the exception and add a getException to it, so there is  not need to have a list inside the listener and the code looks like
 Object value = feel.evaluate(expr, variables);
if (listener.getException() != null) 
   throw listener.getException();

And listener code 

public class FeelReturnValueEvaluatorListener implements FEELEventListener {

    private FeelReturnValueEvaluatorException ;

    @Override
    public void onEvent(FEELEvent event) {
         switch (event.getSeverity()) {
            case ERROR:
                LOG.error("{}", event);
                StringBuilder messageBuilder = new StringBuilder(event.getSeverity().toString()).append(" ").append(event.getMessage());
                if (event.getOffendingSymbol() != null) {
                    messageBuilder.append(" ( offending symbol: '").append(event.getOffendingSymbol()).append("' )");
                }
                if (event.getSourceException() != null) {
                    messageBuilder.append("  ").append(event.getSourceException().getMessage());
                  exception = FeelReturnValueEvaluatorException(messageBuilder.toString(), event.getSourceException()); 
                   break;
                }
               else {
                 exception =  new FeelReturnValueEvaluatorException(messageBuilder.toString());
                }
            case TRACE:
                LOG.debug("{}", event);
                break;
            case WARN:
                LOG.warn("{}", event);
                break;
            case INFO:
            default:
                LOG.info("{}", event);
               break;
    }

    public  getException() {
        return exception;
    }
}

**Comment by elguardian** [comment]
@tarilabs plz take care of @fjtirado comments. I would suggest to go for the full list of errors.

**Review by fjtirado — APPROVED** [review]
Great work @tarilabs 
Thanks a lot for refining the mesage

File: jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluatorListener.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.kie.dmn.api.feel.runtime.events.FEELEvent;
+import org.kie.dmn.api.feel.runtime.events.FEELEventListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FeelReturnValueEvaluatorListener implements FEELEventListener {
```
Comment (danielezonca) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluatorListener.java:27]: Considering that it collects only `ERROR` events, what about rename it as
```suggestion
public class FeelErrorEvaluatorListener implements FEELEventListener {
```

**Review by danielezonca — APPROVED** [review]
LGTM :+1:  minor comment

**Comment by tarilabs** [comment]
>  72.5% Coverage

mainly due to LOG lines uncovered, I would say false positive.

Please consider for merge with squash please

**Comment by elguardian** [comment]
Ok, thx guys. I consider this good to go.

**Pull request merged**